atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCacheImpl.java

```diff
@@ -0,0 +1,163 @@
+import java.util.Comparator;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * The operations on this class *must* be synchronised to protect against concurrent modification of the cache.
+ */
+public final class LockWatchValueCacheImpl implements LockWatchValueCache {
+    private final LockWatchEventCache eventCache;
+    private final ValueStore valueStore;
+    private final SnapshotStore snapshotStore;
+
+    private volatile Optional<LockWatchVersion> currentVersion = Optional.empty();
+
+    public LockWatchValueCacheImpl(LockWatchEventCache eventCache) {
+        this.eventCache = eventCache;
+        this.valueStore = new ValueStoreImpl();
+        this.snapshotStore = new SnapshotStoreImpl();
+    }
+
+    @Override
+    public synchronized void processStartTransactions(Set<Long> startTimestamps) {
+        TransactionsLockWatchUpdate updateForTransactions =
+                eventCache.getUpdateForTransactions(startTimestamps, currentVersion);
+        updateCurrentVersion(updateForTransactions);
+        updateStores(updateForTransactions);
+    }
+
+    // TODO(jshah): This needs to be *very* carefully wired to ensure that the synchronised aspect here is not an
+    //  issue. Chances are that this may need to be re-jigged to take a batch, and be connected to the batched commit
+    //  timestamp call.
+    @Override
+    public synchronized void updateCache(TransactionDigest digest, long startTs) {
+        CommitUpdate commitUpdate = eventCache.getCommitUpdate(startTs);
+        commitUpdate.accept(new Visitor<Void>() {
+            @Override
+            public Void invalidateAll() {
+                // If this is an election, we should throw. If not and we are not a serialisable transaction, we are
+                // *technically* ok to go on here. However, we cannot determine which of the two we are in, so we
+                // throw anyway.
+                throw new TransactionLockWatchFailedException("A Timelock leader election has occurred between the "
+                        + "start and commit time of this transaction, and thus all events have been invalidated. This"
+                        + " transaction will be retried");
+            }
+
+            @Override
+            public Void invalidateSome(Set<LockDescriptor> invalidatedLocks) {
+                Set<CellReference> invalidatedCells = invalidatedLocks.stream()
+                        .flatMap(LockWatchValueCacheImpl::extractTableAndCell)
+                        .collect(Collectors.toSet());
+                KeyedStream.stream(digest.loadedValues())
+                        .filterKeys(cellReference -> !invalidatedCells.contains(cellReference))
+                        .forEach(valueStore::putValue);
+                return null;
+            }
+        });
+        snapshotStore.removeTimestamp(StartTimestamp.of(startTs));
```

Comment:
This seems like it should be in a finally block? (Unless it's removed elsewhere)

Reply:
That does make sense, although to be honest I expect it to throw when there has been a leader election, and so chances are it'd be gone anyway. I'll add it in a finally block since I think that makes the most sense code-wise, though.

Reply:
In fact, `eventCache.getCommitUpdate` can throw if there has been an election, so definitely makes sense to wrap this.